### PR TITLE
add diff and pdiff for conditions

### PR DESCRIPTION
**Pull request by shanhuhai5739**
Compare the latest data points
diff means difference
pdiff means Percentage difference

### CODE REVIEW

File: pkg/services/alerting/conditions/reducer.go [hunk 1]
```diff
@@ -94,6 +94,53 @@ func (s *SimpleReducer) Reduce(series *tsdb.TimeSeries) null.Float {
...
+		// get other points
+		points = points[0:i]
+		for i := len(points) - 1; i >= 0; i-- {
+			if points[i][0].Valid {
+				allNull = false
+				value = first - points[i][0].Float64
+				break
+			}
+		}
+	case "pdiff":
```
Comment (bergquist) [inline pkg/services/alerting/conditions/reducer.go]: Can we rename this percent_diff instead? I think that's more explicit.

**Review by bergquist — CHANGES_REQUESTED** [review]

File: pkg/services/alerting/conditions/reducer.go (hunk 1 above)
Comment (shanhuhai5739) [reply pkg/services/alerting/conditions/reducer.go]: that's ok

**Review by bergquist — APPROVED** [review]

**Comment by bergquist** [comment]
@shanhuhai5739 thank you for contributing! :) and :heart: totoro

**Comment by rehevkor5** [comment]
I can't find info about these two functions in the docs, so for those who don't want to read the code, here's my attempt at a verbal explanation of these two functions. Feedback/corrections are very welcome!

`diff()` finds the most recent point (if any exist) then finds the second most recent point (if any exist), and calculates the difference by calculating `most recent point - second most recent point`.

`percent_diff()` finds the same two points, but runs a different calculation. The value is `abs((most recent point - second most recent point) / second most recent point * 100)`. This can be described as the magnitude (absolute value) of the difference between the two points as a percentage (from 0 to 100) of the value of the second most recent point.

File: pkg/services/alerting/conditions/reducer.go [hunk 2]
```diff
@@ -94,6 +94,53 @@ func (s *SimpleReducer) Reduce(series *tsdb.TimeSeries) null.Float {
...
+		for i = len(points) - 1; i >= 0; i-- {
+			if points[i][0].Valid {
+				allNull = false
+				first = points[i][0].Float64
+				break
+			}
+		}
+		// get other points
+		points = points[0:i]
+		for i := len(points) - 1; i >= 0; i-- {
```
Comment (rehevkor5) [inline pkg/services/alerting/conditions/reducer.go]: I'm not sure how expensive this loop is, but it seems like we can skip it if `allNull` is still `true`. Same applies to the second loop for percent_diff.

**Pull request merged**